Documenting an API standard (#403)
UKHomeOffice/engineering-guidance-and-standards: docs/standards/documenting-an-api.md

---
layout: standard
order: 1
title: Documenting an API
date: 2024-04-09
id: SEGAS-XXX
tags:
  - Software design
  - API Design

> ---

Author: ```suggestion
related:
  sections:
    - title: Related links
      items:
        - text: Minimal documentation set for a product
          href:  /standards/minimal-documentation-set-for-a-product/
---
```

> Documenting an API is an important part of the software design lifecycle. It helps consumers of your API, the wider community and even your own team understand the purpose of endpoints, what and how they work, inputs and outputs etc.

Reviewer: I think we need to go beyond specifying endpoints. We need a full [ICD](https://en.wikipedia.org/wiki/Interface_control_document). So I'd include things like how one on-boards, connects and authenticates with the API.

Author: ```suggestion
Documenting an API is an important part of the software design lifecycle. It helps consumers of your API, the wider community and even your own team understand the purpose of endpoints, what and how they work, inputs and outputs, what the requirements are to access it etc.
```

---

## Requirement(s)

- [You MUST include your API on the relevant registers](#you-must-include-your-api-on-the-relevant-registers)
- [You MUST have appropriate status codes returning from each endpoint](#you-must-have-appropriate-status-codes-returning-from-each-endpoint)
- [You MUST have documentation on inputs and outputs of all endpoints](#you-must-have-documentation-on-inputs-and-outputs-of-all-endpoints)
- [You MUST provide specific feedback in responses](#you-must-provide-specific-feedback-in-responses)
- [You MUST use appropriate nouns for resource names](#you-must-use-appropriate-nouns-for-resource-names)

> 

Author: ```suggestion
- [You MUST document how to connect to your API](#you-must-document-how-to-connect-to-your-api)

```

### You MUST include your API on the relevant registers

Listing our API's contributes to the government community as well as other organisations and wider industry. Listing our API's here will also help inter-departmental data exchanges, as well as sharing best practice in API development.

If your API is for internal or team use only, then it should be included in the internal API register.

> if your API is public, it should also be on the [cabinet offices list of API's](https://www.api.gov.uk/ho/#home-office).

Reviewer: ```suggestion
if your API is public, it should also be on the [cabinet offices list of APIs](https://www.api.gov.uk/ho/#home-office).
```

### You MUST have appropriate status codes returning from each endpoint 

Your API endpoints must return the relevant status code for each endpoint, for example a 403 if a user doesn't have access.

It's important that the correct status codes are used as these are recognisable codes and can make it easier to debug for any consumers.

> The Mozilla Foundation has a [definitive list of status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status), and what they mean and do.

Reviewer: This seems to be more about API design than documentation / specification.

> The Mozilla Foundation has a [definitive list of status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status), and what they mean and do.

Author: Move this to Designing and Maintaining an API. Wrap into inputs and outputs with link to other standard.

### You MUST have documentation on inputs and outputs of all endpoints

> Documenting the inputs and outputs of endpoints makes it easy for consumers to understand what is required and how to use the given responses from an endpoint. It should also detail an error messages that can be returned and reason for it.

Reviewer 2: ```suggestion
Documenting the inputs and outputs of endpoints makes it easy for consumers to understand what is required and how to use the given responses from an endpoint. 

It should include all of the following inputs that apply to your API:

- Request method
- URI query parameters
- Request headers
- Requirements for the request body

For each possible response, including error responses, you should document the following:

- Response status code
- Response headers
- The format of the response body
```

> This can be achieved easily in RESTful API by using Swagger.

Reviewer: I wonder if we should mandate OpenAPI (Swagger v3) for ReSTful APIs. - Are there any serious competitors?

Author: We have another standard around Designing an API, where we think it could be worth including a point around using OpenAPI for new APIs.
```suggestion
This can be achieved for a RESTful API by using [OpenAPI (Swagger v3)](https://www.openapis.org/what-is-openapi).
```

Author: Maybe look at alternative wording to include graphQL

### You MUST provide specific feedback in responses

It's important that response is clear and gives a brief overview of what has happened when called an endpoint. For example. Access Denied or Cannot connect to database.

This can make it easier to debug for any consumers.

A response should not include any information about the technologies used or any internal information for example internal server addresses etc.

### You MUST use an appropriate method

Your API endpoints must use a relevant method for each endpoint, for example, using POST for a login endpoint.

The Mozilla Foundation has a [definitive list of methods](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods), and what they mean and do.

Using the appropriate methods help to keep endpoints secure as well as helping customers understand how an endpoint should be used, how things should be passed and what the end result will be.

### You MUST use appropriate nouns for resource names

Your API resources must follow RESTful resource naming standards as outlined on the restful API [resource naming page](https://restfulapi.net/resource-naming/).

> 

Author: This is dependant on the API (Find ICD link).

> 

Author: Move this to Designing and Maintaining an API standard

> ---

Author: Move appropriate method to Designing and Maintaining an API standard.

---
layout: standard
order: 1
title: Documenting an API
date: 2025-03-21
id: SEGAS-00016
tags:
  - Software design
  - API Design
related:
  sections:
    - title: Related links
      items:
        - text: Minimal documentation set for a product
          href: /standards/minimal-documentation-set-for-a-product/
        - text: How to document APIs
          href: https://www.gov.uk/guidance/how-to-document-apis
        - text: Writing API Reference Documentation

>           href: https://www.gov.uk/guidance/writing-api-reference-documentation

Author: Add designing and maintaining an API.

---

## Requirement(s)

- [You MUST include your API on the relevant registers](#you-must-include-your-api-on-the-relevant-registers)
- [You MUST have documentation on inputs and outputs of all endpoints](#you-must-have-documentation-on-inputs-and-outputs-of-all-endpoints)
- [You MUST provide specific feedback in responses](#you-must-provide-specific-feedback-in-responses)
- [You MUST document how to connect to your API](#you-must-document-how-to-connect-to-your-api)
- [You MUST document how to raise an issue with your API](#you-must-document-how-to-raise-an-issue-with-your-api)
- [You MUST document information on rate limits](#you-must-document-information-on-rate-limits)

### You MUST include your API on the relevant registers if your API is public

Listing our APIs contributes to the government community as well as other organisations and wider industry. Listing our APIs here will also help inter-departmental data exchanges, as well as sharing best practice in API development.

> If your API is public, it should also be on the [UK public sector API Catalogue](https://www.api.gov.uk/ho/#home-office).

Author: ```suggestion
If your API is public, it must also be on the [UK public sector API Catalogue](https://www.api.gov.uk/ho/#home-office).
```

Documenting the requests and responses of endpoints makes it easy for consumers to understand what is required and how to use the given responses from an endpoint.

It should include all of the following that apply to a request:

- Request method
- URI query parameters
- Request headers
- Requirements for the request body

For each possible response, including error responses, you should document the following:

- Response status code
- Response headers
- The format of the response body

This can be achieved for a RESTful API by using [OpenAPI (Swagger v3)](https://www.openapis.org/what-is-openapi).

### You MUST provide specific feedback in error responses

It's important that response is clear and gives a brief overview of what has happened when called an endpoint. For example. Access Denied or Not Found. Your documentation should also state what each status code represents.

This can make it easier to debug for any consumers.

A response should not include any information about the technologies used or any internal information for example internal server addresses etc.

### You MUST document how to connect to your API

> You must have documentation in place explaining how you onboard users to your API, if your API requires authorisation or authentication, you must document how to gain access to the API.

Author: ```suggestion
You must have documentation in place explaining how you onboard users to your API, if your API requires authorisation or authentication, you must document how to gain access to the API.

If you have multiple versions of your API you need document each supported version.
```

> ### You MUST document how to raise an issue with your API

Author: ```suggestion
### You MUST document how to raise an issue with your API if your API is public
```

It's important to document how to raise an issue with your API, this not only helps your consumers with debugging or raising concerns, but enables a feedback loop with your users.

### You MUST document information on rate limits

Rate limiting is important to secure your API and consumers may want to query your API often, so documenting the rate limits helps consumers build their software around these limits.

> 

Author: ```suggestion
### You MUST document different versions of your API
```

Changes, ways to migrate, depreciated properties, breaking changes.

---

## Requirement(s)

- [You MUST include your API on the relevant registers](#you-must-include-your-api-on-the-relevant-registers)
- [You MUST have documentation on inputs and outputs of all endpoints](#you-must-have-documentation-on-inputs-and-outputs-of-all-endpoints)
- [You MUST provide specific feedback in responses](#you-must-provide-specific-feedback-in-responses)
- [You MUST document how to connect to your API](#you-must-document-how-to-connect-to-your-api)
- [You MUST document how to raise an issue with your API](#you-must-document-how-to-raise-an-issue-with-your-api)
- [You MUST document information on rate limits](#you-must-document-information-on-rate-limits)

### You MUST include your API on the relevant registers if your API is public

Listing our APIs contributes to the government community as well as other organisations and wider industry. Listing our APIs here will also help inter-departmental data exchanges, as well as sharing best practice in API development.

> If your API is public, it should also be on the [cabinet offices list of APIs](https://www.api.gov.uk/ho/#home-office).

Reviewer 3: ```suggestion
If your API is public, it should also be on the [UK public sector API Catalogue](https://www.api.gov.uk/ho/#home-office).
```

### You MUST have documentation on inputs and outputs of all endpoints

Documenting the inputs and outputs of endpoints makes it easy for consumers to understand what is required and how to use the given responses from an endpoint.

> It should include all of the following inputs that apply to your API:

Reviewer 3: Rather than inputs/outputs, should we instead use the more widely used request/response terminology?

- Request method
- URI query parameters
- Request headers
- Requirements for the request body

For each possible response, including error responses, you should document the following:

- Response status code
- Response headers
- The format of the response body

This can be achieved for a RESTful API by using [OpenAPI (Swagger v3)](https://www.openapis.org/what-is-openapi).

### You MUST provide specific feedback in responses

> It's important that response is clear and gives a brief overview of what has happened when called an endpoint. For example. Access Denied or Cannot connect to database. Your documentation should also state what each status code represents.

Reviewer 3: Would "Cannot connect to <database>" conflict with the last line in this section?

(For clarity, I agree with don't include information about underlying implementations in API responses)

This can make it easier to debug for any consumers.

> A response should not include any information about the technologies used or any internal information for example internal server addresses etc.

Reviewer 3: The heading of this section mentions specific feedback for responses, but the content only mentions error scenarios. Should the heading be changed to focus on error scenarios, or the content account for happy path scenarios too?

### You MUST document how to connect to your API

> You must have documentation in place explaining how you onboard users to your API, if your API requires authorisation, you must document how to get authentication keys.

Reviewer 3: Check usage of authorisation vs authentication here: https://auth0.com/docs/get-started/identity-fundamentals/authentication-and-authorization

Author: Tried to change this wording as it's more than likely either/both.

### You MUST document how to raise an issue with your API

> It's important to document how to raise an issue with your API, this not only helps your consumers with debugging or raising concerns, but enabled a feedback loop with your users.

Reviewer 3: ```suggestion
It's important to document how to raise an issue with your API, this not only helps your consumers with debugging or raising concerns, but enables a feedback loop with your users.
```

        - text: How to document APIs
          href: https://www.gov.uk/guidance/how-to-document-apis
        - text: Writing API Reference Documentation
          href: https://www.gov.uk/guidance/writing-api-reference-documentation
---

Documenting an API is an important part of the software design lifecycle. It helps consumers of your API, the wider community and even your own team understand the purpose of endpoints, what and how they work, inputs and outputs, what the requirements are to access it etc.

---

## Requirement(s)

- [You MUST include your API on the relevant registers](#you-must-include-your-api-on-the-relevant-registers)
- [You MUST have documentation on inputs and outputs of all endpoints](#you-must-have-documentation-on-inputs-and-outputs-of-all-endpoints)
- [You MUST provide specific feedback in responses](#you-must-provide-specific-feedback-in-responses)
- [You MUST document how to connect to your API](#you-must-document-how-to-connect-to-your-api)
- [You MUST document how to raise an issue with your API](#you-must-document-how-to-raise-an-issue-with-your-api)
- [You MUST document information on rate limits](#you-must-document-information-on-rate-limits)

> ### You MUST include your API on the relevant registers

Author: ```suggestion
### You MUST include your API on the relevant registers if your API is public
```

Listing our APIs contributes to the government community as well as other organisations and wider industry. Listing our APIs here will also help inter-departmental data exchanges, as well as sharing best practice in API development.

> If your API is for internal or team use only, then it should be included in the internal API register.

Author: ```suggestion
```

- Response status code
- Response headers
- The format of the response body

This can be achieved for a RESTful API by using [OpenAPI (Swagger v3)](https://www.openapis.org/what-is-openapi).

### You MUST provide specific feedback in responses

It's important that response is clear and gives a brief overview of what has happened when called an endpoint. For example. Access Denied or Cannot connect to database. Your documentation should also state what each status code represents.

This can make it easier to debug for any consumers.

A response should not include any information about the technologies used or any internal information for example internal server addresses etc.

### You MUST document how to connect to your API

You must have documentation in place explaining how you onboard users to your API, if your API requires authorisation, you must document how to get authentication keys.

### You MUST document how to raise an issue with your API

It's important to document how to raise an issue with your API, this not only helps your consumers with debugging or raising concerns, but enabled a feedback loop with your users.

### You MUST document information on rate limits

> Rate limiting is important to secure your API and consumers may want to query your API often, so documenting the rate limits helps consumers build their software around these limits.

Author: ```suggestion
Rate limiting is important to secure your API and consumers may want to query your API often, so documenting the rate limits helps consumers build their software around these limits. This information may not want to be shared widely if rate limiting is in place for security reasons.
```

Author: Add where appropriate to sharing parts of the documentation more widely.

Author: Add to opening paragraph.

Reviewer 2: If the rate limit being public is enough to make your application insecure, then it's not an adequate security measure. A malicious user can deterimine the limit through trial and error. Publishing the limit is not the issue, and makes your API harder to use for legitimate consumers. 

A project can justify not meeting any part of any standard if their circumstances warrant it. This doesn't need to be a loophole encouraged by the standard.

Author: This was mentioned in the guild meeting @jeff-horton-ho-sas but I do agree with you!

---

## Requirement(s)

- [You MUST include your API on the relevant registers](#you-must-include-your-api-on-the-relevant-registers)
- [You MUST have appropriate status codes returning from each endpoint](#you-must-have-appropriate-status-codes-returning-from-each-endpoint)
- [You MUST have documentation on inputs and outputs of all endpoints](#you-must-have-documentation-on-inputs-and-outputs-of-all-endpoints)
- [You MUST provide specific feedback in responses](#you-must-provide-specific-feedback-in-responses)
- [You MUST use appropriate nouns for resource names](#you-must-use-appropriate-nouns-for-resource-names)
- [You MUST document how to connect to your API](#you-must-document-how-to-connect-to-your-api)

### You MUST include your API on the relevant registers

Listing our APIs contributes to the government community as well as other organisations and wider industry. Listing our APIs here will also help inter-departmental data exchanges, as well as sharing best practice in API development.

> If your API is for internal or team use only, then it should be included in the internal API register.

Author: Add link to internal register

### You MUST have appropriate status codes returning from each endpoint 

Your API endpoints must return the relevant status code for each endpoint, for example a 403 if a user doesn't have access.

It's important that the correct status codes are used as these are recognisable codes and can make it easier to debug for any consumers.

The Mozilla Foundation has a [definitive list of status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status), and what they mean and do.

### You MUST have documentation on inputs and outputs of all endpoints

Documenting the inputs and outputs of endpoints makes it easy for consumers to understand what is required and how to use the given responses from an endpoint. 

It should include all of the following inputs that apply to your API:

- Request method
- URI query parameters
- Request headers
- Requirements for the request body

For each possible response, including error responses, you should document the following:

- Response status code
- Response headers
- The format of the response body

This can be achieved for a RESTful API by using [OpenAPI (Swagger v3)](https://www.openapis.org/what-is-openapi).

> ### You MUST provide specific feedback in responses

Author: The documentation should include what status codes represent.

It's important that response is clear and gives a brief overview of what has happened when called an endpoint. For example. Access Denied or Cannot connect to database.

This can make it easier to debug for any consumers.

A response should not include any information about the technologies used or any internal information for example internal server addresses etc.

> ### You MUST use an appropriate method

Author: The documentation should include what methods do.
